File: python/08.py
#! /usr/bin/env python3

SPACE = '.'

# Load grid into nested array:
with open('../inputs/input08.txt') as fp:
    grid = [[char for char in line_str.strip()] for line_str in fp.readlines()]
ydim = len(grid)
xdim = len(grid[0])

nodes_by_char = {}

# find nodes
for y in range(ydim):
    for x in range(xdim):
        char = grid[y][x]
        if char != SPACE:
            if char not in nodes_by_char:
                nodes_by_char[char] = []
            nodes_by_char[char].append((y,x))


def solve(min_antinode_reach, max_antinode_reach):
    all_antinodes = set()
    for char in nodes_by_char:
        nodes = nodes_by_char[char]
        # consider all combinations of 2 different nodes
        for n1 in nodes:
            for n2 in nodes:
                if n1 == n2:
                    continue
                y1, x1 = n1
                y2, x2 = n2
                delta = (y2 - y1, x2 - x1)
                # reach ahead to find each harmonic antinode
                reach = min_antinode_reach
                while reach < max_antinode_reach:
                    y3 = y2 + (delta[0] * reach)
                    x3 = x2 + (delta[1] * reach)
                    if y3 < 0 or y3 >= ydim or x3 < 0 or x3 >= xdim:
                        # force search to stop in this direction
                        reach = max_antinode_reach
                        break
                    antinode = (y3, x3)
                    all_antinodes.add(antinode)
                    reach += 1

    return len(all_antinodes)


print("Part 1:", solve(1, 2))
print("Part 2:", solve(0, max(ydim, xdim)))

# P1: 214
# P2: 809

File: rust/aoc/src/bin/08.rs
use std::fs::read_to_string;
use std::collections::HashMap;
use std::collections::HashSet;

const SPACE: char = '.';

// https://doc.rust-lang.org/stable/rust-by-example/std_misc/file/read_lines.html
fn read_lines(filename: &str) -> Vec<String> {
    read_to_string(filename)
        .unwrap()
        .lines()
        .map(String::from)
        .collect()
}


fn prep_grid(lines: Vec<String>) -> Vec<Vec<char>> {
    lines
    .iter()
    .map(|line| {
        line.chars().collect::<Vec<char>>()
    })
    .collect::<Vec<Vec<char>>>()
}


fn fill_nodes_by_char(grid: Vec<Vec<char>>) -> HashMap<char, Vec<(i32, i32)>> {
    let ydim = grid.len();
    let xdim = grid[0].len();

    let mut nodes_by_char: HashMap<char, Vec<(i32, i32)>> = HashMap::new();

    // find all the char nodes and group them by char
    for y in 0..ydim {
        for x in 0..xdim {
            let coord = (y as i32, x as i32);
            let c = grid[y][x];

            if c != SPACE {
                nodes_by_char.entry(c)
                    .and_modify(|list| list.push(coord))
                    .or_insert(vec![coord]);
            }
        }
    }

    return nodes_by_char
}


// find the amount of antinodes in the grid
#[allow(unused_assignments)]
fn solve(grid: Vec<Vec<char>>, nodes_by_char: HashMap<char, Vec<(i32, i32)>>, min_antinode_reach: i32, max_antinode_reach: i32) -> i32 {
    let ydim = grid.len() as i32;
    let xdim = grid[0].len() as i32;

    let mut all_antinodes = HashSet::new();

    for (_c, nodes) in &nodes_by_char {
        // consider all combinations of 2 different nodes
        for n1 in nodes {
            for n2 in nodes {
                if n1 == n2 {
                    continue
                }
                let (y1, x1) = n1;
                let (y2, x2) = n2;
                let delta = (y2 - y1, x2 - x1);
                // reach ahead to find each harmonic antinode
                let mut reach = min_antinode_reach;
                while reach < max_antinode_reach {
                    let y3 = y2 + (delta.0 * reach);
                    let x3 = x2 + (delta.1 * reach);
                    if y3 < 0 || y3 >= ydim || x3 < 0 || x3 >= xdim {
                        // force search to stop in this direction
                        reach = max_antinode_reach;
                        break
                    }
                    all_antinodes.insert((y3, x3));
                    reach += 1
                }
            }
        }
    }

    return all_antinodes.len() as i32
}


fn main() {
    let grid = prep_grid(read_lines("../../inputs/input08.txt"));
    let nodes_by_char = fill_nodes_by_char(grid.clone());
    println!("P1: {}", solve(grid.clone(), nodes_by_char.clone(), 1, 2));
    println!("P2: {}", solve(grid.clone(), nodes_by_char.clone(), 0, 50));
}

// P1: 214
// P2: 809